src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,6 +177,21 @@ private bool CompleteCore()
             {
                 if (existingWaiters != null)
                 {
+                    // we have waiters and we have data.
+                    // allow some extra time to drain the queue before cancelling
+                    // but do not wait for too long.
+                    SpinWait.SpinUntil(() =>
+                    {
```

Comment:
AsyncQueue is intended to be a general facility that could be promoted to the BCL. This is placing app-specific policy into its implementation that would be inappropriate in a general-purpose facility. Please move this policy into the caller.